File: packages/butler-agent/src/agent/tools/project-ledger/closeout.ts
import { runProjectLedgerTool } from "../../../integrations/project-ledger/client.ts";
import { projectLedgerNativeNextHints } from "./recovery-hints.ts";

type ProjectLedgerCloseoutInput = {
  butlerHome: string;
  butlerData?: string;
};

type ProjectLedgerToolRunner = typeof runProjectLedgerTool;
type ProjectLedgerViewName = "dashboard" | "handoff" | "roadmap";

const LIFECYCLE_CLOSEOUT_TOOLS = new Set([
  "project_ledger_create",
  "project_ledger_update",
  "project_ledger_work_update",
  "project_ledger_work_complete",
  "project_ledger_task_update",
  "project_ledger_task_complete",
]);

const GENERATED_VIEWS: readonly ProjectLedgerViewName[] = [
  "dashboard",
  "handoff",
  "roadmap",
];

export function needsProjectLedgerLifecycleCloseout(
  toolName: string,
  result: Record<string, unknown>,
): boolean {
  return LIFECYCLE_CLOSEOUT_TOOLS.has(toolName) && result.ok === true;
}

export function runProjectLedgerLifecycleCloseout(input: {
  executor: ProjectLedgerCloseoutInput;
  projectPath: string;
  runTool?: ProjectLedgerToolRunner;
  refreshedIndex?: Record<string, unknown> | null;
}): Record<string, unknown> {
  const runTool = input.runTool ?? runProjectLedgerTool;
  const project = ["--project", input.projectPath];
  const indexResult = input.refreshedIndex ?? runTool(input.executor, ["index", ...project]);
  const indexOk = indexResult.ok === true;
  if (!indexOk) {
    return {
      ok: false,
      index_ok: false,
      index_path: stringAt(indexResult, ["data", "index", "path"]),
      index_error: projectLedgerErrorSummary(indexResult),
      rendered_views: GENERATED_VIEWS.map((view) => skippedViewSummary(view, "index_failed")),
      check_ok: false,
      check_skipped: true,
      issue_count: 0,
      issues: [],
      failed_stages: ["index"],
    };
  }
  const renderedViews = GENERATED_VIEWS.map((view) =>
    renderViewSummary(view, runTool(input.executor, [
      "render",
      ...project,
      view,
      "--write",
    ])),
  );
  const checkResult = runTool(input.executor, ["check", ...project]);
  const renderOk = renderedViews.every((view) => view.ok === true);
  const checkOk = checkResult.ok === true;
  const failedStages = closeoutFailedStages({
    indexOk,
    renderedViews,
    checkOk,
  });
  return {
    ok: failedStages.length === 0,
    index_ok: indexOk,
    index_path: stringAt(indexResult, ["data", "index", "path"]),
    rendered_views: renderedViews,
    check_ok: checkOk,
    issue_count: issueCount(checkResult),
    issues: boundedIssues(checkResult),
    failed_stages: failedStages,
    ...(renderOk ? {} : { render_error: renderErrorSummary(renderedViews) }),
    ...(checkOk ? {} : { check_error: projectLedgerErrorSummary(checkResult) }),
  };
}

export function applyProjectLedgerLifecycleCloseout(
  result: Record<string, unknown>,
  closeout: Record<string, unknown>,
): Record<string, unknown> {
  if (closeout.ok === true) {
    return {
      ...result,
      project_ledger_closeout: closeout,
    };
  }
  return {
    ok: false,
    recoverable: true,
    observation_kind: "validation_failed",
    error: projectLedgerCloseoutFailureError(closeout),
    mutation_result: result,
    project_ledger_closeout: closeout,
  };
}

function renderViewSummary(
  view: ProjectLedgerViewName,
  result: Record<string, unknown>,
): Record<string, unknown> {
  const path = stringAt(result, ["data", "path"]);
  const written = booleanAt(result, ["data", "written"]);
  const ok = result.ok === true && written && Boolean(path);
  return {
    view,
    ok,
    path,
    written,
    ...(ok ? {} : { error: renderFailureSummary(result, written, path) }),
  };
}

function renderFailureSummary(
  result: Record<string, unknown>,
  written: boolean,
  path: string | null,
): Record<string, unknown> {
  if (result.ok !== true) return projectLedgerErrorSummary(result);
  return {
    code: "project_ledger_render_not_written",
    message: !written
      ? "Project Ledger render reported success without writing the generated view."
      : "Project Ledger render reported success without a generated view path.",
    next: [],
    native_next: [{
      tool: "project_ledger_render",
      args: { write: true },
      reason: path
        ? "Rerun Project Ledger render with write enabled and verify the generated view path."
        : "Rerun Project Ledger render with write enabled so the generated view path is available.",
    }],
  };
}

function skippedViewSummary(view: ProjectLedgerViewName, reason: string): Record<string, unknown> {
  return {
    view,
    ok: false,
    path: null,
    written: false,
    skipped: true,
    reason,
  };
}

function closeoutFailedStages(input: {
  indexOk: boolean;
  renderedViews: Array<Record<string, unknown>>;
  checkOk: boolean;
}): string[] {
  const stages: string[] = [];
  if (!input.indexOk) stages.push("index");
  const renderFailed = input.renderedViews.some((view) => view.ok !== true && view.skipped !== true);
  if (renderFailed) stages.push("render");
  if (!input.checkOk) stages.push("check");
  return stages;
}

function renderErrorSummary(renderedViews: Array<Record<string, unknown>>): Record<string, unknown> {
  return {
    code: "project_ledger_render_failed",
    message: "One or more Project Ledger generated views failed to render.",
    details: renderedViews
      .filter((view) => view.ok !== true && view.skipped !== true)
      .map((view) => ({
        view: stringValue(view.view),
        error: view.error,
      })),
  };
}

function issueCount(result: Record<string, unknown>): number {
  const data = objectAt(result, ["data"]);
  const value = data?.issueCount;
  return typeof value === "number" && Number.isFinite(value)
    ? value
    : boundedIssues(result).length;
}

function boundedIssues(result: Record<string, unknown>): Array<Record<string, unknown>> {
  const data = objectAt(result, ["data"]);
  const issues = Array.isArray(data?.issues) ? data.issues : [];
  return issues.slice(0, 8).map((issue) => {
    const item = recordValue(issue);
    return {
      code: stringValue(item.code),
      severity: stringValue(item.severity),
      message: stringValue(item.message),
      path: stringValue(item.path),
      record: stringValue(item.record),
    };
  });
}

function projectLedgerErrorSummary(result: Record<string, unknown>): Record<string, unknown> {
  const error = recordValue(result.error);
  return {
    code: stringValue(error.code),
    message: stringValue(error.message),
    next: boundedNext(error.next),
    native_next: projectLedgerNativeNextHints(error),
  };
}

function projectLedgerCloseoutFailureError(closeout: Record<string, unknown>): Record<string, unknown> {
  return {
    code: "project_ledger_closeout_failed",
    message: "Project Ledger lifecycle closeout failed after a successful mutation.",
    details: closeoutFailureDetails(closeout),
    native_next: closeoutNativeNextHints(closeout),
  };
}

function closeoutFailureDetails(closeout: Record<string, unknown>): Array<Record<string, string>> {
  const details: Array<Record<string, string>> = [];
  if (closeout.index_ok !== true) {
    details.push({
      code: "index_failed",
      kind: "project_ledger_closeout",
      status: "failed",
      message: stringAt(closeout, ["index_error", "message"]) ?? "Project Ledger index failed.",
    });
  }
  const renderedViews = Array.isArray(closeout.rendered_views) ? closeout.rendered_views : [];
  for (const view of renderedViews) {
    const record = recordValue(view);
    if (record.ok === true || record.skipped === true) continue;
    details.push({
      code: "render_failed",
      kind: "project_ledger_closeout",
      id: stringValue(record.view) ?? "view",
      status: "failed",
      message: stringAt(record, ["error", "message"]) ?? "Project Ledger generated view render failed.",
    });
  }
  if (closeout.check_ok !== true && closeout.check_skipped !== true) {
    details.push({
      code: "check_failed",
      kind: "project_ledger_closeout",
      status: "failed",
      message: stringAt(closeout, ["check_error", "message"]) ?? "Project Ledger check failed.",
    });
  }
  return details;
}

function closeoutNativeNextHints(closeout: Record<string, unknown>): Array<Record<string, unknown>> {
  const hints: Array<Record<string, unknown>> = [];
  if (closeout.index_ok !== true) {
    hints.push({
      tool: "project_ledger_index",
      args: {},
      reason: "Repair Project Ledger source records if needed, then rebuild the compact index.",
    });
  }
  const renderedViews = Array.isArray(closeout.rendered_views) ? closeout.rendered_views : [];
  for (const view of renderedViews) {
    const record = recordValue(view);
    if (record.ok === true || record.skipped === true) continue;
    hints.push({
      tool: "project_ledger_render",
      args: { view: stringValue(record.view) ?? "dashboard", write: true },
      reason: "Rewrite the failed generated Project Ledger view after index succeeds.",
    });
  }
  if (closeout.check_ok !== true && closeout.check_skipped !== true) {
    hints.push({
      tool: "project_ledger_check",
      args: {},
      reason: "Review issues, repair source records, and rerun strict validation.",
    });
  }
  return hints;
}

function boundedNext(value: unknown): Array<Record<string, string>> {
  if (!Array.isArray(value)) return [];
  return value.slice(0, 5).map((item) => {
    const record = recordValue(item);
    return {
      command: stringValue(record.command) ?? "",
      reason: stringValue(record.reason) ?? "",
    };
  });
}

function objectAt(
  source: Record<string, unknown>,
  path: readonly string[],
): Record<string, unknown> | null {
  let current: unknown = source;
  for (const key of path) {
    const record = recordValue(current);
    if (!(key in record)) return null;
    current = record[key];
  }
  return recordValue(current);
}

function stringAt(source: Record<string, unknown>, path: readonly string[]): string | null {
  let current: unknown = source;
  for (const key of path) {
    const record = recordValue(current);
    if (!(key in record)) return null;
    current = record[key];
  }
  return typeof current === "string" && current.trim() ? current.trim() : null;
}

function booleanAt(source: Record<string, unknown>, path: readonly string[]): boolean {
  let current: unknown = source;
  for (const key of path) {
    const record = recordValue(current);
    if (!(key in record)) return false;
    current = record[key];
  }
  return current === true;
}

function recordValue(value: unknown): Record<string, unknown> {
  return value && typeof value === "object" && !Array.isArray(value)
    ? value as Record<string, unknown>
    : {};
}

function stringValue(value: unknown): string | null {
  return typeof value === "string" && value.trim() ? value.trim() : null;
}

File: packages/butler-agent/rust/agent/src/host/guided_project_tools/lifecycle/closeout.rs
//! Post-mutation index, persisted views, check, and source closeout projection.

use std::path::Path;

use serde_json::{Value, json};

use crate::project_ledger::{LedgerCommand, NativeProjectLedger, ProjectLedgerReadError};

use super::{command, nonempty};

const VIEWS: [&str; 3] = ["dashboard", "handoff", "roadmap"];

pub(crate) async fn closeout(
    ledger: &NativeProjectLedger,
    root: &Path,
    mutation: Value,
) -> Result<Value, ProjectLedgerReadError> {
    let refreshed = mutation
        .pointer("/data/derived/index_refresh")
        .filter(|value| value.get("ok") == Some(&Value::Bool(true)));
    let index = if let Some(refreshed) = refreshed {
        json!({"ok":true,"data":{"index":{"path":refreshed.get("path")}}})
    } else {
        command(ledger, root, LedgerCommand::Index, json!({})).await?
    };
    let index_ok = index.get("ok") == Some(&Value::Bool(true));
    let index_path = nonempty(index.pointer("/data/index/path"));
    if !index_ok {
        let closeout = json!({
            "ok":false,"index_ok":false,"index_path":index_path,
            "index_error":error_summary(&index),
            "rendered_views":VIEWS.map(|view| json!({"view":view,"ok":false,"path":null,"written":false,"skipped":true,"reason":"index_failed"})),
            "check_ok":false,"check_skipped":true,"issue_count":0,"issues":[],"failed_stages":["index"],
        });
        return Ok(apply_closeout(mutation, closeout));
    }
    let mut rendered = Vec::with_capacity(VIEWS.len());
    for view in VIEWS {
        let result = command(
            ledger,
            root,
            LedgerCommand::Render,
            json!({"view":view,"write":true}),
        )
        .await?;
        let path = nonempty(result.pointer("/data/path"));
        let written = result.pointer("/data/written") == Some(&Value::Bool(true));
        let ok = result.get("ok") == Some(&Value::Bool(true)) && written && path.is_some();
        let mut summary = json!({"view":view,"ok":ok,"path":path,"written":written});
        if !ok {
            summary["error"] = if result.get("ok") == Some(&Value::Bool(true)) {
                let message = if written {
                    "Project Ledger render reported success without a generated view path."
                } else {
                    "Project Ledger render reported success without writing the generated view."
                };
                let reason = if path.is_some() {
                    "Rerun Project Ledger render with write enabled and verify the generated view path."
                } else {
                    "Rerun Project Ledger render with write enabled so the generated view path is available."
                };
                json!({"code":"project_ledger_render_not_written","message":message,"next":[],"native_next":[{"tool":"project_ledger_render","args":{"view":view,"write":true},"reason":reason}]})
            } else {
                error_summary(&result)
            };
        }
        rendered.push(summary);
    }
    let check = command(ledger, root, LedgerCommand::Check, json!({})).await?;
    let check_ok = check.get("ok") == Some(&Value::Bool(true));
    let issues = check
        .pointer("/data/issues")
        .and_then(Value::as_array)
        .map(Vec::as_slice)
        .unwrap_or(&[]);
    let bounded: Vec<_> = issues.iter().take(8).map(|issue| json!({
        "code":nullable_text(issue.get("code")),"severity":nullable_text(issue.get("severity")),
        "message":nullable_text(issue.get("message")),"path":nullable_text(issue.get("path")),
        "record":nullable_text(issue.get("record")),
    })).collect();
    let count = check
        .pointer("/data/issueCount")
        .and_then(Value::as_f64)
        .filter(|number| number.is_finite())
        .map(Value::from)
        .unwrap_or_else(|| json!(bounded.len()));
    let render_failed = rendered
        .iter()
        .any(|view| view.get("ok") != Some(&Value::Bool(true)));
    let mut failed = Vec::new();
    if render_failed {
        failed.push("render");
    }
    if !check_ok {
        failed.push("check");
    }
    let mut closeout = json!({
        "ok":failed.is_empty(),"index_ok":true,"index_path":index_path,
        "rendered_views":rendered,"check_ok":check_ok,"issue_count":count,
        "issues":bounded,"failed_stages":failed,
    });
    if render_failed {
        closeout["render_error"] = json!({"code":"project_ledger_render_failed","message":"One or more Project Ledger generated views failed to render.","details":rendered.iter().filter(|view| view.get("ok") != Some(&Value::Bool(true))).map(|view| json!({"view":view.get("view"),"error":view.get("error")})).collect::<Vec<_>>()});
    }
    if !check_ok {
        closeout["check_error"] = error_summary(&check);
    }
    Ok(apply_closeout(mutation, closeout))
}

fn nullable_text(value: Option<&Value>) -> Value {
    nonempty(value).map_or(Value::Null, |value| value.into())
}

fn error_summary(result: &Value) -> Value {
    let error = result.get("error").unwrap_or(&Value::Null);
    let source_next = error
        .get("next")
        .and_then(Value::as_array)
        .map(Vec::as_slice)
        .unwrap_or(&[]);
    let next: Vec<_> = source_next
        .iter()
        .take(5)
        .map(|item| {
            json!({
                "command":nonempty(item.get("command")).unwrap_or(""),
                "reason":nonempty(item.get("reason")).unwrap_or("")
            })
        })
        .collect();
    let native_next = if source_next.iter().any(|item| {
        item.as_str()
            .or_else(|| item.get("command").and_then(Value::as_str))
            .is_some_and(|command| {
                let command = command.trim();
                command == "project-ledger index"
                    || command.starts_with("project-ledger index ")
                    || command == "pl index"
                    || command.starts_with("pl index ")
            })
    }) {
        vec![
            json!({"tool":"project_ledger_index","args":{},"reason":"Rebuild the compact Project Ledger index for this project."}),
        ]
    } else if matches!(
        error.get("code").and_then(Value::as_str),
        Some("record_not_found" | "ambiguous_record")
    ) {
        vec![
            json!({"tool":"project_ledger_list","args":{"kind":"all"},"reason":"List records, then retry with the exact id and kind."}),
        ]
    } else if error.get("code").and_then(Value::as_str) == Some("project_ledger_check_failed") {
        vec![
            json!({"tool":"project_ledger_check","args":{},"reason":"Review data.issues, repair source records, and rerun validation."}),
        ]
    } else {
        vec![]
    };
    json!({"code":nullable_text(error.get("code")),"message":nullable_text(error.get("message")),"next":next,"native_next":native_next})
}

fn apply_closeout(mut mutation: Value, closeout: Value) -> Value {
    if closeout.get("ok") == Some(&Value::Bool(true)) {
        mutation["project_ledger_closeout"] = closeout;
        return mutation;
    }
    let details = failure_details(&closeout);
    let hints = closeout_hints(&closeout);
    json!({"ok":false,"recoverable":true,"observation_kind":"validation_failed",
        "error":{"code":"project_ledger_closeout_failed","message":"Project Ledger lifecycle closeout failed after a successful mutation.","details":details,"native_next":hints},
        "mutation_result":mutation,"project_ledger_closeout":closeout})
}

fn failure_details(closeout: &Value) -> Vec<Value> {
    let mut details = Vec::new();
    if closeout.get("index_ok") != Some(&Value::Bool(true)) {
        details.push(json!({"code":"index_failed","kind":"project_ledger_closeout","status":"failed","message":nonempty(closeout.pointer("/index_error/message")).unwrap_or("Project Ledger index failed.")}));
    }
    for view in closeout
        .get("rendered_views")
        .and_then(Value::as_array)
        .map(Vec::as_slice)
        .unwrap_or(&[])
    {
        if view.get("ok") == Some(&Value::Bool(true))
            || view.get("skipped") == Some(&Value::Bool(true))
        {
            continue;
        }
        details.push(json!({"code":"render_failed","kind":"project_ledger_closeout","id":nonempty(view.get("view")).unwrap_or("view"),"status":"failed","message":nonempty(view.pointer("/error/message")).unwrap_or("Project Ledger generated view render failed.")}));
    }
    if closeout.get("check_ok") != Some(&Value::Bool(true))
        && closeout.get("check_skipped") != Some(&Value::Bool(true))
    {
        details.push(json!({"code":"check_failed","kind":"project_ledger_closeout","status":"failed","message":nonempty(closeout.pointer("/check_error/message")).unwrap_or("Project Ledger check failed.")}));
    }
    details
}

fn closeout_hints(closeout: &Value) -> Vec<Value> {
    let mut hints = Vec::new();
    if closeout.get("index_ok") != Some(&Value::Bool(true)) {
        hints.push(json!({"tool":"project_ledger_index","args":{},"reason":"Repair Project Ledger source records if needed, then rebuild the compact index."}));
    }
    for view in closeout
        .get("rendered_views")
        .and_then(Value::as_array)
        .map(Vec::as_slice)
        .unwrap_or(&[])
    {
        if view.get("ok") == Some(&Value::Bool(true))
            || view.get("skipped") == Some(&Value::Bool(true))
        {
            continue;
        }
        hints.push(json!({"tool":"project_ledger_render","args":{"view":nonempty(view.get("view")).unwrap_or("dashboard"),"write":true},"reason":"Rewrite the failed generated Project Ledger view after index succeeds."}));
    }
    if closeout.get("check_ok") != Some(&Value::Bool(true))
        && closeout.get("check_skipped") != Some(&Value::Bool(true))
    {
        hints.push(json!({"tool":"project_ledger_check","args":{},"reason":"Review issues, repair source records, and rerun strict validation."}));
    }
    hints
}
